samples/custom_dispatch/cpu/plugin/example.mlir

```diff
@@ -0,0 +1,154 @@
+// CHECK-SYSTEM: mul[7](2 * 4 = 8)
+// arith.addf 8 + 4 = 12
+// CHECK-SYSTEM: 8xf32=12 12 12 12 12 12 12 12
+
+// This example demonstrates calling dynamically imported functions in the
+// runtime. Alternatively the functions can be embedded into the compiled IREE
+// programs for hermetic deployment (see custom_dispatch/cpu/embedded/).
+
+// The configurations used for executable compilation.
+// This lets the compiler and runtime know the format and requirements of the
+// executable binaries produced. For this sample this matches what we expect
+// the plugins to be compiled for (architecture-wise).
+#aarch64_target = #hal.executable.target<"llvm-cpu", "embedded-elf-arm_64", {
+  data_layout = "e-m:e-p270:32:32-p271:32:32-p272:64:64-i64:64-f80:128-n8:16:32:64-S128",
+  native_vector_size = 16 : index,
+  target_triple = "aarch64-unknown-unknown-eabi-elf"
+}>
+#x86_64_target = #hal.executable.target<"llvm-cpu", "embedded-elf-x86_64", {
+  data_layout = "e-m:e-p270:32:32-p271:32:32-p272:64:64-i64:64-f80:128-n8:16:32:64-S128",
+  native_vector_size = 32 : index,
+  target_triple = "x86_64-unknown-unknown-eabi-elf"
+}>
+
+// The target devices that the program will run on.
+// These can come from compiler flags and multiple targets can be supported
+// It's possible, for example, to support targeting multiple devices in the same
+// compiled binary (CPU + Vulkan, etc).
+#cpu_target = #hal.device.target<"llvm-cpu", {
+  executable_targets = [
+    #aarch64_target,
+    #x86_64_target
+  ]
+}>
+
+module @example attributes {hal.device.targets = [#cpu_target]} {
+
+  // Executable containing exported shims and calls to external functions.
+  // Each executable can contain multiple exported functions and variants for
+  // different architectures or even devices. It's also possible to mix hand-
+  // authored functions with code generated ones even for the same functions
+  // such that code generation is used as a fallback when the hand-authored
+  // kernels aren't supported at runtime.
+  stream.executable private @executable {
+    stream.executable.export public @simple_mul workgroups(%workload: index) -> (index, index, index) {
+      // This host function is used to compute the XYZ workgroup count
+      // dispatched at runtime. It can query the %device for capabilities
+      // and limits (last-level cache sizes, etc). The other arguments are the
+      // values passed in the dispatch operation (usually things like root
+      // output op tensor dimensions and other abstract values).
+      %x = affine.apply affine_map<()[s0] -> (s0 ceildiv 64)>()[%workload]
+      %c1 = arith.constant 1 : index
+      stream.return %x, %c1, %c1 : index, index, index
+    }
+
+    builtin.module {
+      // External function declaration using a user-chosen calling convention.
+      func.func private @simple_mul_workgroup(%binding0: memref<?xf32>, %binding1: memref<?xf32>, %binding2: memref<?xf32>, %dim: index, %tid: index)
```

Comment:
Should this have an attribute or use something other than `func.func`? I'd expect external / plugin-defined functions to be easier to spot (control+F to search for across a module, etc.)

Reply:
externally defined functions are just externally defined functions - nothing special about them - lets us use upstream lowerings to func.call and such without requiring IREE-specific internals to leak (especially useful in plugins that can use just upstream dialects)